proof/refine/ARM/SchedContextInv_R.thy

```diff
@@ -1012,6 +1012,268 @@ lemma refillUpdate_corres:
    apply clarsimp+
   done
 
+lemma head_insufficient_length_cross:
+  "\<lbrakk>(s,s') \<in> state_relation; valid_objs s; pspace_aligned s; pspace_distinct s;
+    pred_map (\<lambda>cfg. unat MIN_BUDGET \<le> refills_unat_sum (scrc_refills cfg))
+             (sc_refill_cfgs_of s) scPtr;
+    the (head_insufficient scPtr s); valid_objs' s'\<rbrakk>
+   \<Longrightarrow> obj_at' (\<lambda>sc'. Suc 0 < scRefillCount sc') scPtr s'"
+  apply (frule (1) head_insufficient_length_at_least_two[rotated])
+  apply (clarsimp simp: vs_all_heap_simps)
+  apply (rename_tac sc n)
+  apply (prop_tac "valid_sched_context_size n")
+   apply (fastforce intro: valid_sched_context_size_objsI)
+  apply (prop_tac "sc_at' scPtr s'")
+   apply (fastforce intro!: sc_at_cross
+                      simp: vs_all_heap_simps obj_at_kh_kheap_simps is_sc_obj_def)
+  apply (clarsimp simp: obj_at'_def projectKOs)
+  apply (rename_tac sc')
+  apply (prop_tac "sc_relation sc n sc'")
+   apply (frule state_relation_pspace_relation)
+   apply (clarsimp simp: pspace_relation_def vs_all_heap_simps)
+   apply (drule_tac x=scPtr in bspec, blast)
+   apply clarsimp
+  apply (prop_tac "valid_sched_context' sc' s'")
+   apply (frule sc_ko_at_valid_objs_valid_sc'[rotated])
+    apply (fastforce simp: obj_at'_def projectKOs)
+   apply (fastforce dest!: sc_ko_at_valid_objs_valid_sc')
+  apply (auto simp: sc_relation_def refills_map_def valid_sched_context'_def)
+  done
+
+lemma non_overlapping_merge_refills_rewrite:
+  "update_refill_hd scPtr (r_time_update (\<lambda>t. t - r_amount old_head)
+                           \<circ> r_amount_update (\<lambda>m. m + r_amount old_head))
+   = do update_refill_hd scPtr (r_amount_update (\<lambda>m. m + r_amount old_head));
+        update_refill_hd scPtr (r_time_update (\<lambda>t. t - r_amount old_head))
+     od"
```

Comment:
Can't we do this simply by unfolding `update_refill_hd` and using an existing lemma about `update_sched_context`?

Reply:
I tried that, but I couldn't get it to work. I might take another look at it